File: backend/vention_printer_interface/control/heater_model.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class HeaterExposure:
    powder_mass_g: float
    carbon_mass_g: float
    ink_mass_g: float
    ipa_mass_g: float
    energy_j: float
    time_s: float

    def sweep_speed_mm_s(self, pass_length_mm: float) -> float:
        return pass_length_mm / self.time_s if self.time_s > 0 else 0.0


def exposure(*, layer_mm: float, area_mm2: float, carbon_wt: float, powder_density_g_cm3: float,
             ink_carbon_wt: float, ipa_dhvap_j_g: float, section_power_w: float,
             passes: int = 1) -> HeaterExposure:
    """IPA-only heater exposure (spec §8). Ink is `ink_carbon_wt` carbon / the rest IPA.
    powder_mass = density * volume (per layer); carbon deposited = powder * wt/(1-wt) * `passes`
    (multi-pass stacks `passes` layers of ink evaporated in ONE heating pass); ink = carbon /
    ink_carbon_wt; IPA = ink * (1 - ink_carbon_wt); energy = IPA * dHvap; time = energy / power."""
    volume_cm3 = (area_mm2 * layer_mm) / 1000.0
    powder = powder_density_g_cm3 * volume_cm3
    carbon = (powder * carbon_wt / (1.0 - carbon_wt) if carbon_wt < 1.0 else 0.0) * max(passes, 1)
    ink = carbon / ink_carbon_wt if ink_carbon_wt > 0 else 0.0
    ipa = ink * (1.0 - ink_carbon_wt)
    energy = ipa * ipa_dhvap_j_g
    time = energy / section_power_w if section_power_w > 0 else 0.0
    return HeaterExposure(round(powder, 6), round(carbon, 6), round(ink, 6), round(ipa, 6),
                          round(energy, 3), round(time, 3))

File: frontend/src/lib/heater_model.ts
/** Mirror of backend heater_model.exposure(): IPA-only heater exposure (spec §8).
 *  Ink is `inkCarbonWt` carbon / the rest IPA; only IPA evaporates. Multi-pass jetting
 *  stacks `passes` layers of ink evaporated in ONE heating pass, so energy scales with passes. */

export interface HeaterExposure {
  powderMassG: number;
  carbonMassG: number;
  inkMassG: number;
  ipaMassG: number;
  energyJ: number;
  timeS: number;
  sweepSpeed(passLenMm: number): number;
}

export interface ExposureArgs {
  layerMm: number;
  areaMm2: number;
  carbonWt: number;
  powderDensityGCm3: number;
  inkCarbonWt: number;
  ipaDhvapJG: number;
  sectionPowerW: number;
  passes?: number;
}

const round = (x: number, n: number): number => {
  const f = 10 ** n;
  return Math.round(x * f) / f;
};

export function exposure(args: ExposureArgs): HeaterExposure {
  const { layerMm, areaMm2, carbonWt, powderDensityGCm3, inkCarbonWt, ipaDhvapJG,
    sectionPowerW, passes = 1 } = args;
  const volumeCm3 = (areaMm2 * layerMm) / 1000.0;
  const powder = powderDensityGCm3 * volumeCm3;
  const carbon = (carbonWt < 1.0 ? (powder * carbonWt) / (1.0 - carbonWt) : 0.0) * Math.max(passes, 1);
  const ink = inkCarbonWt > 0 ? carbon / inkCarbonWt : 0.0;
  const ipa = ink * (1.0 - inkCarbonWt);
  const energy = ipa * ipaDhvapJG;
  const time = sectionPowerW > 0 ? energy / sectionPowerW : 0.0;
  const timeS = round(time, 3);
  return {
    powderMassG: round(powder, 6),
    carbonMassG: round(carbon, 6),
    inkMassG: round(ink, 6),
    ipaMassG: round(ipa, 6),
    energyJ: round(energy, 3),
    timeS,
    sweepSpeed(passLenMm: number): number {
      return timeS > 0 ? passLenMm / timeS : 0.0;
    },
  };
}
